### PR TITLE
bind comparator function to object scope

### PR DESCRIPTION
**Pull request by moos**
This allows (deep) sorting based on keys **and** values.

Use case: I have an object representation of a deep folder structure.  This change allows doing breadth-first sort.
Source:
```js
{
  c: 'c'
  a: 'a',
  b: { ... },
}
```
Current sorts keys only:
```js
{
  a: 'a',
  b: { ... },
  c: 'c'
}
```
This PR will allow considering the value as well (in this case, nested objects are sorted last)
```js
{
  a: 'a',
  c: 'c',
  b: { ... },
}
```

### CODE REVIEW

**Comment by sindresorhus** [comment]
Makes sense. Can you add docs and a test?

File: test.js [hunk 1]
```diff
@@ -1,17 +1,36 @@
 import test from 'ava';
 import sortKeys from '.';
 
+const orderedDeepEqual = (t, a, b) => {
+	t.deepEqual(a, b);
+	t.deepEqual(Object.keys(a), Object.keys(b));
+};
```
Comment (moos) [inline test.js]: `t.deepEqual` doesn't verify key order, which is critical in these tests.  So adding second assertion based on the keys.

File: readme.md [hunk 2]
```diff
@@ -32,6 +32,20 @@ sortKeys({c: 0, a: 0, b: 0}, {
...
+  // compare based on keys and values
+  compare: function(left, right) {
+      let lvalue = this[left];
+      let rvalue = this[right];
+      if (lvalue === rvalue) return left.localeCompare(right);
+      return lvalue < rvalue;
+    }
+});
+```
+This can be used, e.g., to group all functions or objects last.  Do not use arrow function here since it would clobber `this`.
```
Comment (sindresorhus) [inline readme.md]: This should be documented in the API section and should not mention it's new behavior. That's for the release notes to mention.

File: package.json [hunk 3]
```diff
@@ -37,5 +41,5 @@
 		"ava": "^2.2.0",
 		"tsd": "^0.7.4",
 		"xo": "^0.24.0"
-	}
+    }
```
Comment (sindresorhus) [inline package.json]: Don't do unrelated changes.

File: package.json [hunk 4]
```diff
@@ -9,7 +9,11 @@
 		"email": "sindresorhus@gmail.com",
 		"url": "sindresorhus.com"
 	},
-	"engines": {
+    "contributors": [{
+      "name": "Moos",
+      "email": "mooster@42at.com"
+    }],
```
Comment (sindresorhus) [inline package.json]: Thanks, but I don't use this field. It's annoying having to keep it updated. The commit history speaks for itself.

File: readme.md [hunk 5]
```diff
@@ -32,6 +32,20 @@ sortKeys({c: 0, a: 0, b: 0}, {
...
 ```
 
+As of v5.x, the comparator function, if given, will be bound to the current object.  It means `this` is available to access the object's values, e.g.:
+
+```js
+sortkeys({{c: 0, a: 0, b: 1}}, {
+  // compare based on keys and values
+  compare: function(left, right) {
+      let lvalue = this[left];
+      let rvalue = this[right];
```
Comment (sindresorhus) [inline readme.md]: Looking at the example here now, I'm realizing that it would be nicer to just passing in the object as the third parameter instead. That would make it work with arrow functions and be more explicit. I don't really like implicit this binding in general.

File: readme.md [hunk 6]
```diff
@@ -32,6 +32,20 @@ sortKeys({c: 0, a: 0, b: 0}, {
...
+      if (lvalue === rvalue) return left.localeCompare(right);
+      return lvalue < rvalue;
+    }
```
Comment (sindresorhus) [inline readme.md]: Use tab-indentation.

File: test.js [hunk 7]
```diff
@@ -1,17 +1,36 @@
...
 });
 
 test('custom compare function', t => {
-	const compare = (a, b) => a.localeCompare(b);
-	t.deepEqual(sortKeys({c: 0, a: 0, b: 0}, {compare}), {c: 0, b: 0, a: 0});
+	const compare = (a, b) => -a.localeCompare(b);
+	orderedDeepEqual(t, sortKeys({c: 0, a: 0, b: 0}, {compare}), {c: 0, b: 0, a: 0});
+});
+
+test('custom compare function with context', t => {
```
Comment (sindresorhus) [inline test.js]: Also need a test for it in combination with the `deep` option.

File: package.json [hunk 8]
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "sort-keys",
-	"version": "4.0.0",
+	"version": "5.0.0",
```
Comment (sindresorhus) [inline package.json]: Don't bump version in a PR. It's up to the maintainers to choose what version and when to update it.

**Comment by sindresorhus** [comment]
You also need to update https://github.com/sindresorhus/sort-keys/blob/master/index.d.ts

File: test.js (hunk 7 above)
Comment (moos) [reply test.js]: 👍

**Comment by moos** [comment]
Done with changes.

File: readme.md [hunk 9]
```diff
@@ -57,10 +56,11 @@ Recursively sort keys, including keys of objects inside arrays.
 ##### compare
 
 Type: `Function`
+Default: [Compare function.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort)
 
-[Compare function.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort)
-
+If given, the function signature `(a, b, context)` gets a third argument which is the current context, so that the values can be used for comparison as well, as `context[a]` and `context[b]`.
```
Comment (sindresorhus) [inline readme.md]: I don't think "context" is a useful term here. It's the current object.

File: index.d.ts [hunk 10]
```diff
@@ -10,7 +10,7 @@ declare namespace sortKeys {
 		/**
 		[Compare function.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort)
 		*/
-		readonly compare?: (left: string, right: string) => number;
+		readonly compare?: (left: string, right: string, context?: object) => number;
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
		readonly compare?: (left: string, right: string, context?: unknown) => number;
```

Since we cannot know the type.

Comment (sindresorhus) [reply index.d.ts]: The `context` parameter should be documented. See: https://github.com/sindresorhus/typescript-definition-style-guide

**Comment by moos** [comment]
Ping!

File: index.d.ts [hunk 11]
```diff
@@ -9,8 +9,12 @@ declare namespace sortKeys {
 
 		/**
 		[Compare function.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort)
+		
+		@param left - Left key.
```
Comment (sindresorhus) [inline index.d.ts]: Could use a better description.

File: index.d.ts [hunk 12]
```diff
@@ -9,8 +9,12 @@ declare namespace sortKeys {
...
+		@param right - Right key.
+		@param object - The current object whose keys are being compared.
 		*/
-		readonly compare?: (left: string, right: string) => number;
+		readonly compare?: (left: string, right: string, object?: unknown) => number;
```
Comment (sindresorhus) [inline index.d.ts]: Why is `object` marked as optional?

Comment (sindresorhus) [inline index.d.ts]:
```suggestion
		readonly compare?: (leftKey: string, rightKey: string, object?: {[key: string]: unknown}) => number;
```

File: readme.md [hunk 13]
```diff
@@ -57,10 +56,11 @@ Recursively sort keys, including keys of objects inside arrays.
 ##### compare
 
 Type: `Function`
+Default: [Compare function.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort)
 
-[Compare function.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort)
-
+If given, the function signature `(a, b, object)` gets a third argument which is the current object, so that the values can be used for comparison as well, as `object[a]` and `object[b]`.
```
Comment (sindresorhus) [inline readme.md]: This description should be in the TS file too.

Comment (sindresorhus) [inline readme.md]:
```suggestion
If given, the function signature `(a, b, object)` receives a third argument which is the current object, so that the values can be used for the comparison as well, as `object[a]` and `object[b]`.
```

**Comment by sindresorhus** [comment]
Closing for lack of response.
